README.md
[![Build Status](https://travis-ci.org/everyleaf/everyleaf-embulk_helper.svg?branch=master)](https://travis-ci.org/everyleaf/everyleaf-embulk_helper)
[![Gem Version](https://badge.fury.io/rb/everyleaf-embulk_helper.svg)](http://badge.fury.io/rb/everyleaf-embulk_helper)
[![Code Climate](https://codeclimate.com/github/everyleaf/everyleaf-embulk_helper/badges/gpa.svg)](https://codeclimate.com/github/everyleaf/everyleaf-embulk_helper)
[![Test Coverage](https://codeclimate.com/github/everyleaf/everyleaf-embulk_helper/badges/coverage.svg)](https://codeclimate.com/github/everyleaf/everyleaf-embulk_helper/coverage)

# Everyleaf::EmbulkHelper



## Installation

Add this line to your application's Gemfile:

```ruby
gem 'everyleaf-embulk_helper'
```

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install everyleaf-embulk_helper

Load rake tasks at your Rakefile:

    require "bundler/gem_tasks"
    require "everyleaf/embulk_helper/tasks"

    Everyleaf::EmbulkHelper::Tasks.install({
      gemspec: "./embulk-input-your-plugin.gemspec",
      github_name: "uu59/embulk-input-your-plugin",
    })

## Usage

    $ bundle exec rake -T
    rake build                     # Build embulk-input-your-plugin-0.1.0.gem into the pkg directory
    rake generate:bump_version     # Bump version
    rake generate:changelog        # Generate CHANGELOG.md from previous release
    rake generate:gemfiles         # Generate gemfiles to test this plugin with released Embulk versions (since MIN_VERSION)
    rake generate:prepare_release  # Generate chengelog then bump version
    rake generate:travis           # Generate .travis.yml with gemfiles
    rake install                   # Build and install embulk-input-your-plugin-0.1.0.gem into system gems
    rake release                   # Create tag v0.1.0 and build and push embulk-input-your-plugin-0.1.0.gem to Rubygems

### generate:gemfiles

    $ mkdir gemfiles
    $ cat > gemfiles/template.erb
    source 'https://rubygems.org/'
    gemspec :path => '../'

    gem "embulk", "<%= version %>"

    $ tree gemfiles
    gemfiles
    └── template.erb

    0 directories, 1 file
    $ bundle exec rake generate:gemfiles MIN_VERSION=0.6.10
    I, [2015-08-11T11:03:37.202083 #10238]  INFO -- : Generate Embulk gemfiles from '0.6.10' to latest
    I, [2015-08-11T11:03:38.966539 #10238]  INFO -- : Updated Gemfiles '0.6.10' to '0.6.21'
    $ tree gemfiles
    gemfiles
    ├── embulk-0.6.10
    ├── embulk-0.6.11
    ├── embulk-0.6.12
    ├── embulk-0.6.13
    ├── embulk-0.6.14
    ├── embulk-0.6.15
    ├── embulk-0.6.16
    ├── embulk-0.6.17
    ├── embulk-0.6.18
    ├── embulk-0.6.19
    ├── embulk-0.6.20
    ├── embulk-0.6.21
    ├── embulk-latest
    └── template.erb
    $ cat gemfiles/embulk-latest
    source 'https://rubygems.org/'
    gemspec :path => '../'

    gem "embulk", "> 0.6.10"

    $ cat gemfiles/embulk-0.6.18
    source 'https://rubygems.org/'
    gemspec :path => '../'

    gem "embulk", "0.6.18"

### generate:travis

NOTE: `versions` in template is assigned by file globbing from `gemfiles/*`, thus run `rake generate:gemfiles` before `rake generate:travis`.

    $ cat > .travis.yml.erb
    gemfiles:
    <% versions.each do |v| -%>
      - gemfiles/<%= v %>
    <% end -%>

    $ bundle exec rake generate:travis
    I, [2015-08-11T11:06:53.375018 #10911]  INFO -- : Updated .travis.yml
    $ cat .travis.yml
    gemfiles:
      - gemfiles/embulk-0.6.10
      - gemfiles/embulk-0.6.11
      - gemfiles/embulk-0.6.12
      - gemfiles/embulk-0.6.13
      - gemfiles/embulk-0.6.14
      - gemfiles/embulk-0.6.15
      - gemfiles/embulk-0.6.16
      - gemfiles/embulk-0.6.17
      - gemfiles/embulk-0.6.18
      - gemfiles/embulk-0.6.19
      - gemfiles/embulk-0.6.20
      - gemfiles/embulk-0.6.21
      - gemfiles/embulk-latest

## Contributing

1. Fork it ( https://github.com/[my-github-username]/everyleaf-embulk_helper/fork )
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create a new Pull Request


CHANGELOG.md
## 0.0.6 - 2015-09-09

* [fix] [gemfile] Fix 0.x-latest gemfile [#6](https://github.com/everyleaf/everyleaf-embulk_helper/pull/6)

## 0.0.5 - 2015-09-08

* [improve] [travis] Modify template to test Embulk 0.7.x matrix [#5](https://github.com/everyleaf/everyleaf-embulk_helper/pull/5)

## 0.0.4 - 2015-08-19
* [improve] Ignore errors on -latest versions [#4](https://github.com/everyleaf/everyleaf-embulk_helper/pull/4)

## 0.0.3 - 2015-08-18

* [improve] travis: Initial travis.yml template has allow_failures Embulk 0.6.22 [#3](https://github.com/everyleaf/everyleaf-embulk_helper/pull/3)
* [improve] gemfiles: Genarates embulk-0.x-latest gemfile [#3](https://github.com/everyleaf/everyleaf-embulk_helper/pull/3)

## 0.0.2 - 2015-08-11

* [fix] Fix gemspec path within generated gemfile template [#2](https://github.com/everyleaf/everyleaf-embulk_helper/pull/2)
* [improve] generage:changelog can accept `CURRENT_VER` and `NEXT_VER` environment variable instead of auto detection

## 0.0.1 - 2015-08-11

The first release!!



test/run-test.rb
#!/usr/bin/env ruby

base_dir = File.expand_path(File.join(File.dirname(__FILE__), ".."))
lib_dir = File.join(base_dir, "lib")
test_dir = File.join(base_dir, "test")

require "test-unit"
require "test/unit/rr"
require "codeclimate-test-reporter"
require "pry"

$LOAD_PATH.unshift(lib_dir)
$LOAD_PATH.unshift(test_dir)

ENV["TEST_UNIT_MAX_DIFF_TARGET_STRING_SIZE"] ||= "5000"

CodeClimate::TestReporter.start

require "everyleaf/embulk_helper/tasks"

exit Test::Unit::AutoRunner.run(true, test_dir)


test/everyleaf/embulk_helper/tasks/generator/test_travis.rb
module Everyleaf
  module EmbulkHelper
    module Tasks
      module Generator
        class TestTravis < Test::Unit::TestCase
          def setup
            FileUtils.touch(gemspec)

            @task = Travis.new(options)
            mute_logger(@task)
            File.open(template_path, "w") do |f|
              f.write <<-ERB
<% versions.each do |v| -%>
  - <%= v %>
<% end -%>
              ERB
            end
            stub(@task).embulk_versions { embulk_versions }
          end

          def test_task_installed
            Everyleaf::EmbulkHelper::Tasks.install(options)
            travis_task = Rake::Task.tasks.find do |task|
              task.name == "generate:travis"
            end
            assert travis_task
          ensure
            Rake::Task.clear
          end

          def test_travis_yml_create
            stub(@task).gemfiles { %w(/tmp/foo /tmp/bar).map{|f| Pathname.new(f) } }
            @task.travis
            file = @task.root_dir.join(".travis.yml")

            assert_equal(<<-YML, file.read)
  - foo
  - bar
YML
          end

          private

          def gemspec
            "/tmp/foo.gemspec"
          end

          def template_path
            Pathname.new("/tmp/travis.erb")
          end

          def options
            {
              gemspec: gemspec,
              travis_yml_template: template_path.to_s,
              github_name: "everyleaf/everyleaf-embulk_helper",
            }
          end

          def mute_logger(task)
            stub(task).logger { ::Logger.new(File::NULL) }
          end
        end
      end
    end
  end
end


test/everyleaf/embulk_helper/tasks/generator/test_gemfiles.rb
module Everyleaf
  module EmbulkHelper
    module Tasks
      module Generator
        class TestGemfiles < Test::Unit::TestCase
          def setup
            FileUtils.touch(gemspec)

            @task = Gemfiles.new(options)
            mute_logger(@task)
            template = Pathname.new("/tmp/gem.erb")
            stub(@task).gemfile_template_path { template }
            File.open(template, "w") do |f|
              f.write <<-ERB
gem "embulk", "<%= version %>"
              ERB
            end
            stub(@task).embulk_versions { embulk_versions }
          end

          class TestMinVersion < self
            def test_without_specified
              @task.gemfiles
              files = Dir["#{@task.root_dir}/gemfiles/embulk-*"]

              # all versions + minor versions + latest version
              assert_equal(embulk_versions.length + minor_versions.length + 1, files.length)
            end

            def test_0_1_1_with_env
              ENV["MIN_VERSION"] = "0.1.1"
              @task.gemfiles
              files = Dir["#{@task.root_dir}/gemfiles/embulk-*"]

              # 0.1.1 + 0.1.2 + 0.1-latest + 0.2.0 + 0.2-latest + latest version
              assert_equal(6, files.length)
            ensure
              ENV.delete("MIN_VERSION")
            end

            def test_0_1_1_with_options
              stub(@task).options { options.merge(min_version: "0.1.1") }

              @task.gemfiles
              files = Dir["#{@task.root_dir}/gemfiles/embulk-*"]

              # 0.1.1 + 0.1.2 + 0.1-latest + 0.2.0 + 0.2-latest + latest version
              assert_equal(6, files.length)
            end

            def minor_versions
              @task.send(:target_versions_without_patch)
            end
          end

          class TestContent < self
            def test_content_latest
              @task.gemfiles
              content = @task.root_dir.join("gemfiles/embulk-latest").read
              assert_equal(<<-TXT, content)
gem "embulk", "> 0.0.1"
              TXT
            end

            def test_content_0_1_2
              @task.gemfiles
              content = @task.root_dir.join("gemfiles/embulk-0.1.2").read
              assert_equal(<<-TXT, content)
gem "embulk", "0.1.2"
              TXT
            end
          end

          class TestInitialContent < self
            def setup
              FileUtils.mkdir_p root
              FileUtils.touch gemspec_path

              task = Gemfiles.new(options.merge({
                github_name: "dummy/dummy",
                gemspec: gemspec_path,
                gemfile_template_path: gemfile_template_path,
              }))
              mute_logger(task)
              task.send(:init)
            end

            def teardown
              FileUtils.rm_rf root
            end

            def test_relative_path
              content = File.read(gemfile_template_path)
              assert content.include?("path => '../'") # the gemspec path from each gemfiles/* as relative
            end

            def test_embulk_version_contain
              content = File.read(gemfile_template_path)
              assert content.include?(%Q|gem "embulk", "<%= version %>"|)
            end

            private

            def gemfile_template_path
              root.join("gemfiles/template.erb")
            end

            def root
              Pathname.new("/tmp/foo")
            end

            def gemspec_path
              root.join("bar.spec")
            end
          end

          def test_task_installed
            Everyleaf::EmbulkHelper::Tasks.install(options)
            gemfiles_task = Rake::Task.tasks.find do |task|
              task.name == "generate:gemfiles"
            end
            assert gemfiles_task
          ensure
            Rake::Task.clear
          end

          def test_gemfiles_create
            @task.gemfiles
            files = Dir["#{@task.root_dir}/gemfiles/embulk-*"]

            assert_false(files.empty?)
          end

          private

          def gemspec
            "/tmp/foo.gemspec"
          end

          def embulk_versions
            %W(0.1.0 0.1.1 0.1.2 0.2.0).map{|v| Gem::Version.new(v)}
          end

          def options
            {
              gemspec: gemspec,
              github_name: "everyleaf/everyleaf-embulk_helper",
            }
          end

          def mute_logger(task)
            stub(task).logger { ::Logger.new(File::NULL) }
          end
        end
      end
    end
  end
end


lib/everyleaf/embulk_helper/tasks/common.rb
require "pathname"
require "open-uri"
require "logger"
require "rake"
require "erb"
require "json"
require "rubygems"

module Everyleaf
  module EmbulkHelper
    module Tasks
      module Common
        class OptionError < StandardError; end

        include Rake::DSL

        attr_reader :options

        def initialize(options = {})
          raise "options should be a Hash, given #{options.class}: #{options} " unless options.is_a?(Hash)
          @options = options
          validate_options
        end

        def self.included(klass)
          Tasks.register(klass)
        end

        def install_tasks
          raise NotImplementedError
        end

        def validate_options
          raise OptionError, "gemspec file path is required" unless options[:gemspec]
          raise OptionError, "gemspec file '#{options[:gemspec]}' is not found" unless File.exists?(options[:gemspec])

          required_options.map(&:to_sym).each do |opt|
            raise OptionError, "#{opt} is required" unless options[opt]
          end
        end

        def logger
          ::Logger.new(STDERR)
        end

        def gemspec_path
          Pathname.new(File.expand_path(options[:gemspec]))
        end

        def root_dir
          gemspec_path.dirname
        end

        def gemfiles_dir
          root_dir.join("gemfiles")
        end

        def required_options
          [] # Implement this in subclass if needed
        end

        def embulk_tags
          @embulk_tags ||= JSON.parse(open("https://api.github.com/repos/embulk/embulk/tags").read)
        end

        def embulk_versions
          embulk_tags.map{|tag| Gem::Version.new(tag["name"][/v(.*)/, 1])}.sort
        end
      end
    end
  end
end


lib/everyleaf/embulk_helper/tasks/generator/travis.rb
require "everyleaf/embulk_helper/tasks/common"

module Everyleaf
  module EmbulkHelper
    module Tasks
      module Generator
        class Travis
          include Tasks::Common

          DEFAULT_TRAVIS_YML_TEMPLATE = ".travis.yml.erb".freeze

          def install_tasks
            namespace :generate do
              desc "Generate .travis.yml with gemfiles"
              task :travis do
                travis
              end
            end
          end

          def travis
            init
            create_travis_yml
            logger.info "Updated .travis.yml"
          end

          private

          def init
            FileUtils.mkdir_p File.dirname(travis_yml_template_path)
            unless File.exists?(travis_yml_template_path)
              logger.info "Generate travis.yml template (#{travis_yml_template_path})"
              File.open(travis_yml_template_path, "w") do |f|
                f.write initial_template
              end
            end
          end

          def initial_template
            <<-YML
language: ruby

jdk: oraclejdk8

rvm:
  - jruby-19mode
  - jruby-9.0.0.0

gemfile:
<% versions.each do |file| -%>
  - gemfiles/<%= file %>
<% end -%>

matrix:
  exclude:
    - jdk: oraclejdk8 # Ignore all matrix at first, use `include` to allow build
  include:
    <% matrix.each do |m| -%>
<%= m %>
    <% end %>

  allow_failures:
    - gemfile: gemfiles/embulk-0.6.22
    - gemfile: gemfiles/embulk-0.7.0
    - gemfile: gemfiles/embulk-0.7.1
    # Ignore failure for *-latest
    <% versions.find_all{|file| file.to_s.match(/-latest/)}.each do |file| -%>
- gemfile: <%= file %>
    <% end %>
            YML
          end

          def travis_yml_template_path
            root_dir.join(options[:travis_yml_template] || DEFAULT_TRAVIS_YML_TEMPLATE)
          end

          def create_travis_yml
            erb = ERB.new(travis_yml_template_path.read, nil, "-")
            File.open(root_dir.join(".travis.yml"), "w") do |f|
              f.puts erb.result(binding())
            end
          end

          def gemfiles
            Pathname.glob(gemfiles_dir.join("embulk-*"))
          end

          def versions
            # for trabis.yml.erb
            gemfiles.map(&:basename)
          end

          def matrix
            # for trabis.yml.erb
            gemfiles.map do |path|
              rvm = path.to_s.include?("0.6") ? "jruby-19mode" : "jruby-9.0.0.0"
              %Q|- {rvm: #{rvm}, gemfile: #{path.relative_path_from(root_dir)}}|
            end
          end
        end
      end
    end
  end
end


lib/everyleaf/embulk_helper/tasks/generator/changelog.rb
require "everyleaf/embulk_helper/tasks/common"

module Everyleaf
  module EmbulkHelper
    module Tasks
      module Generator
        class Changelog
          include Tasks::Common

          DEFAULT_CHANGELOG_TEMPLATE = "changelog.erb".freeze

          def install_tasks
            namespace :generate do
              desc "Generate chengelog then bump version"
              task :prepare_release => [:changelog, :bump_version]

              desc "Generate CHANGELOG.md from previous release"
              task :changelog do
                changelog
              end

              desc "Bump version. UP=major to do major version up, UP=minor, UP=patch(default) so on."
              task :bump_version do
                bump_version
                update_gemfile_lock
              end
            end
          end

          def changelog
            content = new_changelog
            File.open(changelog_path, "w") do |f|
              f.write content
            end
          end

          def bump_version
            logger.info "Version bump '#{current_version}' to '#{next_version}'"
            old_content = gemspec_path.read
            new_content = old_content.gsub(/(spec\.version += *)".*?"/, %Q!\\1"#{next_version}"!)
            File.open(gemspec_path, "w") do |f|
              f.write new_content
            end
          end

          private

          def required_options
            %w(github_name)
          end

          def current_version
            ENV["CURRENT_VER"] || Gem::Version.new(gemspec_path.read[/spec\.version += *"([0-9]+\.[0-9]+\.[0-9]+)"/, 1])
          end

          def next_version
            return ENV["NEXT_VER"] if ENV["NEXT_VER"]
            major, minor, patch = current_version.segments
            ver = case version_target
            when "patch"
              [major, minor, patch + 1].join(".")
            when "minor"
              [major, minor + 1, 0].join(".")
            when "major"
              [major + 1, 0, 0].join(".")
            end

            Gem::Version.new(ver)
          end

          def version_target
            ENV["UP"] || options[:version_target] || "patch"
          end

          def update_gemfile_lock
            system("bundle install")
          end

          def github_name
            options[:github_name]
          end

          def pull_request_numbers
            sync_git_repo
            `git log v#{current_version}..origin/master --oneline`.scan(/#[0-9]+/).map do |num_with_hash|
              num_with_hash[/[0-9]+/]
            end
          end

          def pull_request_info(number)
            body = open("https://api.github.com/repos/#{github_name}/issues/#{number}").read
            JSON.parse(body)
          end

          def changes
            pull_request_numbers.map do |number|
              payload = pull_request_info(number)
              "* [] #{payload["title"]} [##{number}](https://github.com/#{github_name}/pull/#{number})"
            end
          end

          def new_changelog
            <<-HEADER
## #{next_version} - #{Time.now.strftime("%Y-%m-%d")}
#{changes.join("\n")}

#{changelog_path.read.chomp}
            HEADER
          end

          def sync_git_repo
            system('git fetch --all')
          end

          def changelog_path
            root_dir.join("CHANGELOG.md")
          end
        end
      end
    end
  end
end


lib/everyleaf/embulk_helper/tasks/generator/gemfiles.rb
require "everyleaf/embulk_helper/tasks/common"

module Everyleaf
  module EmbulkHelper
    module Tasks
      module Generator
        class Gemfiles
          include Tasks::Common

          DEFAULT_EMBULK_GEMFILE_TEMPLATE = "gemfiles/template.erb".freeze

          def install_tasks
            namespace :generate do
              desc "Generate gemfiles to test this plugin with released Embulk versions (since MIN_VERSION)"
              task :gemfiles do
                gemfiles
              end
            end
          end

          def gemfiles
            logger.info "Generate Embulk gemfiles from '#{min_version}' to latest"
            init
            create_gemfiles
            logger.info "Updated Gemfiles '#{min_version}' to '#{embulk_versions.max}'"
          end

          private

          def init
            FileUtils.mkdir_p File.dirname(gemfile_template_path)
            unless File.exists?(gemfile_template_path)
              logger.info "Generate gemfiles template file (#{gemfile_template_path})"
              File.open(gemfile_template_path, "w") do |f|
                f.write initial_template
              end
            end
          end

          def initial_template
            <<-ERB
source 'https://rubygems.org/'
gemspec :path => '#{gemspec_path.dirname.relative_path_from(gemfiles_dir)}/'

gem "embulk", "<%= version %>"
            ERB
          end

          def create_gemfiles
            FileUtils.mkdir_p(gemfiles_dir)

            Dir[gemfiles_dir.join("embulk-*")].each{|f| File.unlink(f)}

            target_versions.each do |version|
              create_gemfile(version)
            end

            # e.g. embulk-0.6-latest
            target_versions_without_patch.each do |version|
              create_gemfile("~> #{version}", "#{version}-latest")
            end

            # embulk-latest
            create_gemfile("> #{min_version}", "latest")
          end

          def create_gemfile(version, name = nil)
            erb = ERB.new(gemfile_template_path.read)
            File.open(gemfiles_dir.join("embulk-#{name || version}"), "w") do |f|
              f.write erb.result(binding())
            end
          end

          def min_version
            Gem::Version.new(ENV["MIN_VERSION"] || options[:min_version] || "0.0.1")
          end

          def gemfile_template_path
            root_dir.join(options[:gemfile_template] || DEFAULT_EMBULK_GEMFILE_TEMPLATE)
          end

          def target_versions
            embulk_versions.find_all do |version|
              version >= min_version
            end
          end

          def target_versions_without_patch
            target_versions.map do |version|
              major, minor, _ = version.segments
              Gem::Version.new([major, minor, 0].join(".")) # NOTE: "~> 0.1.0" -> 0.1.x (< 0.2) / "~> 0.1" -> 0.x (< 1.0)
            end.uniq
          end
        end
      end
    end
  end
end
